File: sources/scalac/transformer/matching/SequenceMatcher.java
package scalac.transformer.matching ;

import scalac.*;
import scalac.ast.Tree;
import scalac.typechecker.*;
import Tree.*;

//import scala.compiler.printer.TextTreePrinter ; // DEBUGGING\
    //import scala.compiler.printer.XMLAutomPrinter ; // DEBUGGING\

import scalac.transformer.TransMatch.Matcher ;
import scalac.ast.* ;
import scalac.symtab.* ;

import java.util.* ;

import ch.epfl.lamp.util.Position;

/** constructs a matcher for a sequence pattern. plays two roles in
 *  described in design pattern Builder.
 *  is the "Director" for "Builder" class BerrySethi.
 *  is the "Builder" for "Director" class TransMatch.
 */

public class SequenceMatcher extends PatternTool {

      CodeFactory cf;

      Matcher _m;

      Tree pat[];
      Tree body[];

    /** translates the det/switching automaton to scala code
     *  precondition: pat.type() corresponds to element type
     */
      Tree addBinderToBody( Tree pat, Tree body ) {

            Type elementType = cf.getElemType_Sequence( pat.type() );

            BindingBerrySethi build = new BindingBerrySethi();
            NondetWordAutom left =  build.automatonFrom( pat, new Integer(0) );
            NondetWordAutom right = build.revnfa;

            //  - - -> left

            DetWordAutom dLeft  =
                  new DetWordAutom( left );

            Matcher mL = new Matcher( _m.owner, _m.selector, null );

            LeftTracerInScala ltis =
                  new LeftTracerInScala( dLeft, elementType, mL, cf);

            Tree stms[] = ltis.getTrace();

            Tree sel = gen.Ident( Position.FIRSTPOS, ltis.resultSym );

            // <- - - right

            DetWordAutom dRight =
                  new DetWordAutom( right, left, dLeft);

            //System.out.println( "vars: "+vars );
            int j;

            final RightTracerInScala rtis =
                  new RightTracerInScala( dRight, left, mL,
                                          cf, pat, elementType );

            Tree stms2[] = rtis.getStms( sel );             // to scala

            // run them, bind

            Tree items[] = new Tree[ stms.length
                                     + stms2.length
                                     + 1 ];

            System.arraycopy( stms, 0, items, 0, stms.length );
            System.arraycopy( stms2, 0, items, stms.length, stms2.length );

            j = stms.length + stms2.length ;
            items[ stms.length + stms2.length ] = body;

            Transformer treeCloner = new Transformer(unit.global) {
                    public Tree transform(Tree tree) {
                        tree = super.transform(tree);
                        if (tree.hasSymbol()) {
                            Object symbol = rtis.helpMap2.get(tree.symbol());
                            if (symbol != null) tree.setSymbol((Symbol)symbol);
                        }
                        return tree;
                    }
                };
            //System.out.println("helpmap");
            //System.out.println( rtis.helpMap2 );

            items[ stms.length + stms2.length ] = treeCloner.transform( body );
            for( Iterator it = rtis.helpMap2.entrySet().iterator(); it.hasNext(); ) {
                Map.Entry e = (Map.Entry)it.next();
                Symbol key = (Symbol)e.getKey();
                Symbol val = (Symbol)e.getValue();
                val.setInfo( key.type() ); // !!! ?
            }

            return gen.mkBlock( body.pos, items );
      }

      /** turns body `case pat(x,y,z) => body' into
       **  `{ <runLeft>;<runRight>;body }' if
       *  necessary.
       *  @param body the array containing the bodies of the visitor
       */
      Tree[] addBindersToBodies( Tree body[] ) {
            //System.out.println("addBindersToBodies");
            Tree nbody[] = new Tree[ body.length ];
            for( int i = 0; i < body.length; i++ ) {
                  Integer iI = new Integer( i );

                  if( !CollectVariableTraverser.containsBinding( pat[ i ] ) )
                        {
                              nbody[ i ] = body[ i ]; // no need for binding
                        }
                  else
                        {
                              nbody[ i ] =
                                    addBinderToBody( pat[ i ], body[ i ] );
                        }
            }
            return nbody;
      }
      Type elementType ;

      /** constructs a word recognizer from an array of patterns which
       *  should all be SequencePatterns ( no wildcard * )
       *  precondition: pat.type corresponds to element type
       *  @param _m          Matcher object, holds the result
       *  @param pat         the (Sequence) patterns
       *  @param body        the bodies
       *  @param defaultCase code that is run when nothing matches. may be null, it
       *                     becomes a ThrowMatchError then

       */
      public void construct( Matcher _m,
                             Tree[] pat,
                             Tree[] body,
                             Tree defaultCase,
                             boolean doBinding ) {
	  //System.err.println("SequenceMatcher::construct");
            this.pat  = pat;
            this.body = body;
            assert body.length == pat.length;
            this._m = _m;

            this.cf = new CodeFactory( unit, infer, _m.pos );

            Type seqType = pat[ 0 ].type();

            elementType = cf.getElemType_Sequence( seqType );

            NondetWordAutom manyNfa[] = new NondetWordAutom[ pat.length ];

            BerrySethi build = new BerrySethi();

            for( int i = 0; i < pat.length; i++ )
                  {
                        manyNfa[ i ] = build.automatonFrom( pat[ i ],
                                                            new Integer( i ));
			//manyNfa[ i ].print();
                  }

            // merge nfas into one if necessary

            NondetWordAutom nfa =
                  (pat.length > 1) ? new NondetWordAutom( manyNfa )
                  : manyNfa[ 0 ];

	    //nfa.print();

            DetWordAutom dfa = new DetWordAutom( nfa );

            WordAutomInScala scalaAut = new WordAutomInScala( dfa,
                                                              elementType,
                                                              _m.owner,
                                                              cf,
							      unit.global.target == Global.TARGET_JVM );
            scalaAut.translate();

            if( defaultCase == null )
                  defaultCase = cf.ThrowMatchError( Position.FIRSTPOS, _m.resultType );

            Tree newbody[] = doBinding ? addBindersToBodies( body ): body;

            // FIXME - CODE FOR TESTING TUPLE STUFF

            _m.tree = scalaAut.getMatcherSwitch( _m.selector,
						 defaultCase,
						 newbody,
						 _m.resultType );
      } // construct (Matcher, Tree[], Tree[], Tree, boolean )

      /** constructor, invoked  by AlgebraicMatcher
       */
      SequenceMatcher( Unit unit, Infer infer ) {
            super( unit, infer );
            //Symbol predefSym = defs.getModule( defs.SCALA, Names.Predef );
            //Scope predefScope = predefSym.members();
      }
} // class SequenceMatcher

File: sources/scala/tools/scalac/transformer/matching/SequenceMatcher.scala


import scalac._;
import scalac.ast.Tree;
import Tree._;

    //import scala.compiler.printer.XMLAutomPrinter ; // DEBUGGING\

import scalac.ast._ ;
import scalac.symtab._ ;

import java.util._ ;

import scala.tools.util.Position;

package scala.tools.scalac.transformer.matching {
/** constructs a matcher for a sequence pattern. plays two roles in
 *  described in design pattern Builder.
 *  is the "Director" for "Builder" class BerrySethi.
 *  is the "Builder" for "Director" class TransMatch.
 */

class SequenceMatcher(unit: CompilationUnit) extends PatternTool(unit) {

//  Console.println("CONSTR SEQUENCEMATCHER");
    final val IGNORED = new Integer(42);

    var cf: CodeFactory = _;
    var _m: PartialMatcher = _;

    var bbuild: BindingBerrySethi  = null;

    /** collects variables
     * @return all variables bound by this binding nfa
     */
    def collectNfaVariables(nfa: NondetWordAutom): Set = {
      val seqVars = new HashSet();
      var j = 0;
      while(j < nfa.nstates) {
        if( nfa.qbinders( j ) != null )
          seqVars.addAll( nfa.qbinders( j ) );
        j = j + 1
      }
      seqVars;
    }


    /** translates the det/switching automaton to scala code
     *  precondition: pat.type() corresponds to element type
     */
    def addBinderToBody( pat:Tree , body:Tree  ): Tree = {
      if( bbuild == null )
        bbuild = new BindingBerrySethi(unit);

      val elementType = cf.getElemType_Sequence( pat.getType() );

        // (a) build *binding* nfa (sequential machine)
      val left =  bbuild.automatonFrom( pat, IGNORED );
      val right = bbuild.revnfa;

        //  (b) determinize + translate L

      val dLeft  = new DetWordAutom( left );

      val ltis = new LeftTracerInScala( dLeft, elementType, _m.owner, cf, _m.selector);

      val trace = ltis.getTrace();

        //  (c) determinize + translate R

      val dRight = new DetWordAutom( right, left, dLeft );

      val seqVars = collectNfaVariables( left );
        //System.out.println("seqVars here are:"+seqVars);
      val rtis = new RightTracerInScala( dRight, seqVars, _m.owner,
                                        cf, pat, elementType );

      // !!! Tree stms2 = rtis.getStms( theTrace, unit, body );
      // !!! gen.mkBlock_( body.pos, trace, stms2 );
      val stms2 = rtis.getStms( trace, unit, body );
      stms2;
    }

    private def buildNfas( pat:Array[Tree]  ): Array[NondetWordAutom] = {
      val build   = new BerrySethi(unit);
      val manyNfa = new Array[NondetWordAutom]( pat.length );
      var  i = 0;
      while( i < pat.length ) {
        manyNfa( i ) = build.automatonFrom( pat( i ),
                                           new Integer( i ));
        i = i + 1;
        //manyNfa[ i ].print();
      }
      manyNfa;
    }

    /** constructs a word recognizer from an array of patterns which
     *  should all be SequencePatterns ( no wildcard * )
     *  precondition: pat.type corresponds to element type
     *  @param _m          Matcher object, holds the result
     *  @param pat         the (Sequence) patterns
     *  @param body        the bodies
     *  @param defaultCase code that is run when nothing matches. may be null, it
     *                     becomes a ThrowMatchError then
     *  @param doBinding   flasg that indicates whether variables should be bound
    */
  def construct(_m: PartialMatcher,  pat: Array[Tree] , body: Array[Tree] , defcase1: Tree, doBinding: Boolean  ): Unit = {
    var defaultCase = defcase1;
    this._m = _m;
    //assert body.length == pat.length;
    if( defaultCase == null )
      defaultCase = cf.ThrowMatchError( cf.pos, _m.resultType );

    this.cf = new CodeFactory( unit, _m.pos );

    val seqType = pat( 0 ).getType();
    val elementType = cf.getElemType_Sequence( seqType );

    // STEP 1 - build nfas for each pattern

    val manyNfa = buildNfas( pat );

    // STEP 2 - recognizing

    // (a) merge nfas into one if necessary
    val nfa = if(pat.length > 1) new NondetWordAutom( manyNfa )  else manyNfa( 0 );

    //nfa.print();

    // (b) determinize
    val dfa = new DetWordAutom( nfa );

    // (c) translate to scala code
    val scalaAut =
      new WordAutomInScala( dfa,
                           elementType,
                           _m.owner,
                           cf,
                           unit.global.target == Global.TARGET_JVM );
    scalaAut.translate();

    // STEP 3 - binding

    var newbody:Array[Tree] = _;
    if( !doBinding )
      newbody = body;
    else {  // this is done in the body of the matching case
      newbody = new Array[Tree](body.length);
      var i = 0;
      while(i < body.length) {
        if( !containsBinding( pat( i ) ) )
          newbody( i ) = body( i ); // no need for binding
        else
          newbody( i ) = addBinderToBody( pat( i ), body( i ) );
        i = i + 1;
      }
    }
    _m.tree = scalaAut.getMatcherSwitch( _m.selector,
                                        defaultCase,
                                        newbody,
                                        _m.resultType );
  } // construct (PartialMatcher, Tree[], Tree[], Tree, boolean )

  } // class SequenceMatcher
  }
